File: app/frontend/src/types/finance.ts
export type FinanceTransactionType = 'income' | 'expense'

export interface FinanceTransaction {
  id: string
  occurred_on: string
  transaction_type: FinanceTransactionType
  category: string
  amount_cents: number
  payment_method: string | null
  vendor_customer: string | null
  description: string | null
  lead_id: string | null
  created_by: string | null
  created_at: string
  updated_at: string
}

export type FinanceTransactionInput = {
  occurred_on: string
  transaction_type: FinanceTransactionType
  category: string
  amount_cents: number
  payment_method?: string | null
  vendor_customer?: string | null
  description?: string | null
  lead_id?: string | null
}

export type FinanceTransactionPatch = Partial<FinanceTransactionInput>

export interface FinanceCategorySummary {
  category: string
  income_cents: number
  expense_cents: number
  net_cents: number
}

export interface FinanceSummary {
  income_cents: number
  expense_cents: number
  net_cents: number
  transaction_count: number
  categories: FinanceCategorySummary[]
}

export type FinanceFilters = {
  start_date?: string
  end_date?: string
  transaction_type?: FinanceTransactionType | ''
}

File: app/backend/app/schemas/finance.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

TransactionType = Literal["income", "expense"]


class FinanceTransactionBase(BaseModel):
    city_id: str | None = None
    occurred_on: date
    transaction_type: TransactionType
    category: str = Field(min_length=1, max_length=80)
    amount_cents: int = Field(gt=0)
    payment_method: str | None = Field(default=None, max_length=80)
    vendor_customer: str | None = Field(default=None, max_length=120)
    description: str | None = None
    lead_id: str | None = None


class FinanceTransactionCreate(FinanceTransactionBase):
    pass


class FinanceTransactionPatch(BaseModel):
    city_id: str | None = None
    occurred_on: date | None = None
    transaction_type: TransactionType | None = None
    category: str | None = Field(default=None, min_length=1, max_length=80)
    amount_cents: int | None = Field(default=None, gt=0)
    payment_method: str | None = Field(default=None, max_length=80)
    vendor_customer: str | None = Field(default=None, max_length=120)
    description: str | None = None
    lead_id: str | None = None


class FinanceTransactionOut(FinanceTransactionBase):
    id: str
    city_id: str
    city_name: str | None = None
    city_slug: str | None = None
    created_by: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class FinanceCategorySummary(BaseModel):
    category: str
    income_cents: int = 0
    expense_cents: int = 0
    net_cents: int = 0


class FinanceSummary(BaseModel):
    income_cents: int = 0
    expense_cents: int = 0
    net_cents: int = 0
    transaction_count: int = 0
    categories: list[FinanceCategorySummary] = Field(default_factory=list)
